File: OpenBurnBarCore/Sources/OpenBurnBarAnalytics/AnalyticsEvent.swift
import Foundation

/// The canonical event registry — the single source of macOS + iOS/widget/keyboard
/// event names, mirroring docs/analytics/event-taxonomy.md. Call sites reference cases,
/// so an off-taxonomy name can't be invented; `rawValue` is the wire name.
///
/// Tier 1 names are SHARED across every platform and are byte-identical to the
/// macOS `AnalyticsEvent`. Tier 2 names here are the iOS trio's platform-specific
/// additions (surface.object.action, snake_case props) — the widget and keyboard
/// surfaces emit outcomes only (never content / keystrokes / text).
public enum AnalyticsEvent: String, CaseIterable, Sendable {

    // MARK: Tier 1 — core cross-platform spine (identical schema on every platform)
    case appSessionStarted = "app.session.started"
    case appSessionEnded = "app.session.ended"
    case appForegrounded = "app.foregrounded"
    case appBackgrounded = "app.backgrounded"
    case appStartupFailed = "app.startup.failed"
    case screenViewed = "screen.viewed"
    case navRouteChanged = "nav.route.changed"
    case authSignInCompleted = "auth.sign_in.completed"
    case authSignUpCompleted = "auth.sign_up.completed"
    case authSignedOut = "auth.signed_out"
    case authAccountDeleted = "auth.account.deleted"
    case onboardingStarted = "onboarding.started"
    case onboardingStepViewed = "onboarding.step.viewed"
    case onboardingCompleted = "onboarding.completed"
    case onboardingDismissed = "onboarding.dismissed"
    case subscriptionUpgradeInitiated = "subscription.upgrade.initiated"
    case settingsChanged = "settings.changed"
    case errorHandled = "error.handled"
    case consentAnalyticsGranted = "consent.analytics.granted"

    // MARK: Tier 2 — dashboard (shared with macOS; iOS reuses the parameterized set)
    case dashboardTimeRangeChanged = "dashboard.time_range.changed"
    case dashboardUnitToggled = "dashboard.unit.toggled"
    case dashboardLaneCardOpened = "dashboard.lane_card.opened"
    case dashboardSessionOpened = "dashboard.session.opened"
    case dashboardRefreshTriggered = "dashboard.refresh.triggered"

    // MARK: Tier 2 — chat (shared family; mobile assistants reuse the schema)
    case chatMessageSent = "chat.message.sent"
    case chatGenerationCompleted = "chat.generation.completed"
    case chatGenerationCancelled = "chat.generation.cancelled"
    case chatGenerationFailed = "chat.generation.failed"
    case chatBackendSwitched = "chat.backend.switched"
    case chatModelSelected = "chat.model.selected"
    case chatAttachmentAdded = "chat.attachment.added"

    // MARK: Tier 2 — quota / budget (shared families)
    case quotaRefreshSucceeded = "quota.refresh.succeeded"
    case quotaRefreshFailed = "quota.refresh.failed"
    case budgetRuleChanged = "budget.rule.changed"

    // MARK: Tier 2 — iOS-specific (platform: ios / ipados)
    /// A primary tab was selected on the mobile tab/sidebar root.
    case mobileTabSelected = "mobile.tab.selected"
    /// The iOS share/connect handoff to a paired Mac was initiated.
    case mobilePairingInitiated = "mobile.pairing.initiated"

    // MARK: Tier 2 — widget (platform: widget) — render/tap OUTCOMES only, no content
    /// A WidgetKit timeline render completed (family + freshness outcome).
    case widgetRendered = "widget.render.completed"
    /// The user tapped the widget (deep-link target only; no rendered values).
    case widgetTapped = "widget.tap.opened"

    // MARK: Tier 2 — keyboard (platform: keyboard) — usage OUTCOME only, NEVER text
    /// The custom keyboard became the active input view (appearance outcome).
    case keyboardActivated = "keyboard.session.activated"
    /// A saved snippet was inserted from the keyboard composer (count bucket only;
    /// never the snippet text, never keystrokes, never the document context).
    case keyboardSnippetInserted = "keyboard.snippet.inserted"

    // MARK: Tier 2 — macOS platform-specific additions
    /// Emitted by the TelemetryService fan-out for an existing privacy-preserving
    /// feature-usage record (feature id + outcome + bucketed duration).
    case featureUsed = "feature.used"
    case dashboardScanRun = "dashboard.scan.run"
    case dashboardRecountRun = "dashboard.recount.run"
    case chatToolInvoked = "chat.tool.invoked"
    case chatPersonaSelected = "chat.persona.selected"
    case chatAttachmentFailed = "chat.attachment.failed"
    case chatHistoryCleared = "chat.history.cleared"
    case chatPanelAction = "chat.panel.action"
    case chatDesktopControlGranted = "chat.desktop_control.granted"
    case chatSearchPerformed = "chat.search.performed"
    case insightsCanvasCreated = "insights.canvas.created"
    case insightsCanvasSelected = "insights.canvas.selected"
    case insightsCanvasDeleted = "insights.canvas.deleted"
    case insightsWidgetChanged = "insights.widget.changed"
    case insightsAnalysisRequested = "insights.analysis.requested"
    case insightsAnalysisCompleted = "insights.analysis.completed"
    case insightsPrivacyModeToggled = "insights.privacy_mode.toggled"
    case insightsVerdictRefreshed = "insights.verdict.refreshed"
    case insightsAuditlogCleared = "insights.auditlog.cleared"
    case quotaRefreshStarted = "quota.refresh.started"
    case quotaSetupSaved = "quota.setup.saved"
    case quotaWorkspaceFilterChanged = "quota.workspace.filter_changed"
    case budgetThresholdWarning = "budget.threshold.warning"
    case budgetThresholdBlocked = "budget.threshold.blocked"
    case cloudsyncCompleted = "cloudsync.completed"
    case cloudsyncFailed = "cloudsync.failed"
    case cloudsyncManualBackupRun = "cloudsync.manual_backup.run"
    case menubarPopoverShown = "menubar.popover.shown"
    case menubarAction = "menubar.action"
    case missionConsoleOpened = "mission_console.opened"
    case wallpaperToggled = "wallpaper.toggled"
    case wallpaperConfigChanged = "wallpaper.config.changed"
    /// Privacy-preserving surface selection for the visual capture toggle. Params:
    /// `provider` (persistedToken), `surface` (cli_pty|desktop_app), `trigger`
    /// (settings|session_header|mobile), `fallback_used` (Bool), `is_eligible` (Bool).
    /// Never includes window titles, bundle IDs beyond persistedToken, or pixel hashes.
    case visualCaptureSurfaceSelected = "visual_capture.surface_selected"

    /// Amplitude category metadata for this event (set via the SDK, never
    /// embedded in the name). Default is `primaryAction`.
    public var category: AnalyticsCategory {
        switch self {
        case .appSessionStarted, .appSessionEnded, .appForegrounded, .appBackgrounded,
             .authSignedOut, .onboardingStarted, .onboardingDismissed, .consentAnalyticsGranted,
             .chatGenerationCompleted, .quotaRefreshStarted, .quotaRefreshSucceeded, .keyboardActivated:
            return .lifecycle
        case .appStartupFailed, .errorHandled, .chatGenerationFailed, .chatAttachmentFailed,
             .quotaRefreshFailed, .budgetThresholdBlocked, .cloudsyncFailed:
            return .error
        case .screenViewed, .navRouteChanged, .onboardingStepViewed,
             .mobileTabSelected, .widgetRendered,
             .menubarPopoverShown, .missionConsoleOpened:
            return .screenView
        case .authSignInCompleted, .authSignUpCompleted, .authAccountDeleted,
             .onboardingCompleted, .subscriptionUpgradeInitiated,
             .chatDesktopControlGranted, .cloudsyncCompleted:
            return .conversionAuth
        default:
            return .primaryAction
        }
    }
}

/// Amplitude event-category metadata. Mirrors the five categories in the taxonomy.
/// Byte-identical to the macOS `AnalyticsCategory`.
public enum AnalyticsCategory: String, Sendable {
    case lifecycle
    case screenView = "screen_view"
    case primaryAction = "primary_action"
    case conversionAuth = "conversion_auth"
    case error
}

File: android/app/src/main/java/com/openburnbar/analytics/AnalyticsEvent.kt
package com.openburnbar.analytics

/**
 * The canonical Android event registry — the single source of Android event
 * names, mirroring docs/analytics/event-taxonomy.md and
 * AgentLens/Services/Analytics/AnalyticsEvent.swift. Call sites reference enum
 * entries, so an off-taxonomy name can't be invented; [wire] is the on-the-wire
 * name and is validated against the taxonomy regex in the unit tests.
 *
 * Tier 1 entries are shared verbatim across every platform. Tier 2 (Android)
 * entries are this platform's surface-specific events; they follow
 * `surface.object.action` with snake_case props and are returned to the
 * taxonomy orchestrator for central merge.
 */
enum class AnalyticsEvent(val wire: String, val category: AnalyticsCategory) {

    // ── Tier 1 — core cross-platform spine ──
    APP_SESSION_STARTED("app.session.started", AnalyticsCategory.LIFECYCLE),
    APP_SESSION_ENDED("app.session.ended", AnalyticsCategory.LIFECYCLE),
    APP_FOREGROUNDED("app.foregrounded", AnalyticsCategory.LIFECYCLE),
    APP_BACKGROUNDED("app.backgrounded", AnalyticsCategory.LIFECYCLE),
    APP_STARTUP_FAILED("app.startup.failed", AnalyticsCategory.ERROR),
    SCREEN_VIEWED("screen.viewed", AnalyticsCategory.SCREEN_VIEW),
    NAV_ROUTE_CHANGED("nav.route.changed", AnalyticsCategory.SCREEN_VIEW),
    AUTH_SIGN_IN_COMPLETED("auth.sign_in.completed", AnalyticsCategory.CONVERSION_AUTH),
    AUTH_SIGN_UP_COMPLETED("auth.sign_up.completed", AnalyticsCategory.CONVERSION_AUTH),
    AUTH_SIGNED_OUT("auth.signed_out", AnalyticsCategory.LIFECYCLE),
    AUTH_ACCOUNT_DELETED("auth.account.deleted", AnalyticsCategory.CONVERSION_AUTH),
    SETTINGS_CHANGED("settings.changed", AnalyticsCategory.PRIMARY_ACTION),
    ERROR_HANDLED("error.handled", AnalyticsCategory.ERROR),
    CONSENT_ANALYTICS_GRANTED("consent.analytics.granted", AnalyticsCategory.LIFECYCLE),
    FEATURE_USED("feature.used", AnalyticsCategory.PRIMARY_ACTION),

    // CMO acquisition funnel (taxonomy-legal aliases of page_viewed / app_opened / …)
    PAGE_VIEWED("page.viewed", AnalyticsCategory.SCREEN_VIEW),
    APP_OPENED("app.opened", AnalyticsCategory.LIFECYCLE),
    CTA_CLICKED("cta.clicked", AnalyticsCategory.CONVERSION_AUTH),
    DOWNLOAD_CLICKED("download.clicked", AnalyticsCategory.CONVERSION_AUTH),
    INSTALL_STARTED("install.started", AnalyticsCategory.LIFECYCLE),
    EMAIL_CAPTURED("email.captured", AnalyticsCategory.CONVERSION_AUTH),

    // ── Tier 1 — subscription / upgrade funnel (shared) ──
    SUBSCRIPTION_UPGRADE_INITIATED("subscription.upgrade.initiated", AnalyticsCategory.CONVERSION_AUTH),

    // ── Tier 2 — Android: chat / assistants (surface: chat*) ──
    CHAT_MESSAGE_SENT("chat.message.sent", AnalyticsCategory.PRIMARY_ACTION),
    CHAT_GENERATION_COMPLETED("chat.generation.completed", AnalyticsCategory.LIFECYCLE),
    CHAT_GENERATION_FAILED("chat.generation.failed", AnalyticsCategory.ERROR),
    CHAT_BACKEND_SWITCHED("chat.backend.switched", AnalyticsCategory.PRIMARY_ACTION),

    // ── Tier 2 — Android: dashboard / pulse (surface: dashboard*) ──
    DASHBOARD_TIME_RANGE_CHANGED("dashboard.time_range.changed", AnalyticsCategory.PRIMARY_ACTION),
    DASHBOARD_LANE_CARD_OPENED("dashboard.lane_card.opened", AnalyticsCategory.PRIMARY_ACTION),

    // ── Tier 2 — Android: quota (surface: quota) ──
    QUOTA_REFRESH_STARTED("quota.refresh.started", AnalyticsCategory.LIFECYCLE),
    QUOTA_REFRESH_SUCCEEDED("quota.refresh.succeeded", AnalyticsCategory.LIFECYCLE),
    QUOTA_REFRESH_FAILED("quota.refresh.failed", AnalyticsCategory.ERROR),

    // ── Tier 2 — Android: store / billing (surface: account) ──
    STORE_PURCHASE_COMPLETED("store.purchase.completed", AnalyticsCategory.CONVERSION_AUTH),

    // ── Tier 2 — Android: search (surface: chat / streams) ──
    // Reuses the shared chat Tier 2 event so the search funnel reads across
    // platforms; the taxonomy's Android section lists chat.search.performed.
    CHAT_SEARCH_PERFORMED("chat.search.performed", AnalyticsCategory.PRIMARY_ACTION),
    ;

    companion object {
        /** The set of valid wire names — the generated allow-list the taxonomy governance requires. */
        val wireNames: Set<String> = entries.map { it.wire }.toSet()
    }
}

/** Amplitude event-category metadata. Mirrors the five categories in the taxonomy. */
enum class AnalyticsCategory(val wire: String) {
    LIFECYCLE("lifecycle"),
    SCREEN_VIEW("screen_view"),
    PRIMARY_ACTION("primary_action"),
    CONVERSION_AUTH("conversion_auth"),
    ERROR("error"),
}
